### ISSUE TITLE
[DOC] Mounting a XML file

### ISSUE THREAD

**Issue by Kochise** [ElektraInitiative/libelektra#4335]
## Steps to Reproduce the Problem

1- Download `note.xml` from https://www.w3schools.com/xml/xml_examples.asp
2- kdb mount -W note.xml user:/sw/myorg/toto/#0/current/
3- kdb ls user:/sw/myorg/toto/#0/current/
4- kdb get sw/myorg/toto/#0/current/note/heading

## Expected Result

user:/sw/myorg/toto/#0/current/note/to
user:/sw/myorg/toto/#0/current/note/from
user:/sw/myorg/toto/#0/current/note/heading
user:/sw/myorg/toto/#0/current/note/body
Reminder

## Actual Result

\<nothing\>

## System Information

- Elektra: 0.9.9

**Comment by markus2330** [comment]
Thank you for reporting! Probably the note.xml was not at the right place? See https://www.libelektra.org/plugins/resolver how resolving of filenames in different namespaces work. You can always check by using `kdb file`.

Following works for me:

```sh
wget https://www.w3schools.com/xml/note.xml
kdb mount -W `pwd`/note.xml system:/sw/myorg/toto/#0/current/ xml
kdb ls system:/sw/myorg/toto/#0/current/
kdb get /sw/myorg/toto/#0/current/heading
kdb set system:/sw/myorg/toto/#0/current/heading No-Reminder
```

I get the output:
```
system:/sw/myorg/toto/#0/current/body
system:/sw/myorg/toto/#0/current/from
system:/sw/myorg/toto/#0/current/heading
system:/sw/myorg/toto/#0/current/to
```

And the modified XML file:
```
<?xml version="1.0" encoding="UTF-8" standalone="no" ?>
<note>

  <body>Don't forget me this weekend!</body>

  <from>Jani</from>

  <heading>No-Reminder</heading>

  <to>Tove</to>

</note>
```

**Comment by Kochise** [comment]
Thank for the clarification.

Should the `kdb mount` command fail and return an error code to indicate something went wrong ?

I used `echo $?` but it returned 0.

**Comment by Kochise** [comment]
```bash
 BINDINGS                         ALL                                                               
 BUILD_DOCSET                     OFF                                                               
 BUILD_DOCUMENTATION              ON                                                                
 BUILD_FULL                       OFF                                                               
 BUILD_PDF                        OFF                                                               
 BUILD_SHARED                     ON                                                                
 BUILD_STATIC                     OFF                                                               
 BUILD_TESTING                    ON                                                                
 CMAKE_BUILD_TYPE                 RelWithDebInfo                                                    
 CMAKE_INSTALL_PREFIX             /usr/local                                                        
 ENABLE_ASAN                      OFF                                                               
 ENABLE_COVERAGE                  OFF                                                               
 ENABLE_DEBUG                     OFF                                                               
 ENABLE_KDB_TESTING               ON                                                                
 ENABLE_LOGGER                    OFF                                                               
 ENABLE_OPTIMIZATIONS             ON                                                                
 ENABLE_TESTING                   ON                                                                
 INSTALL_BUILD_TOOLS              OFF                                                               
 INSTALL_DOCUMENTATION            ON                                                                
 INSTALL_SYSTEM_FILES             OFF                                                               
 INSTALL_TESTING                  ON                                                                
 KDB_DB_DIR                       .dir                                                              
 KDB_DB_FILE                      default.ecf                                                       
 KDB_DB_HOME                      /home                                                             
 KDB_DB_INIT                      elektra.ecf                                                       
 KDB_DB_SPEC                      share/elektra/specification                                       
 KDB_DB_SYSTEM                    /etc/kdb                                                          
 KDB_DB_USER                      .config                                                           
 KDB_DEFAULT_RESOLVER             resolver_fm_hpu_b                                                 
 KDB_DEFAULT_STORAGE              dump                                                              
 LIB_SUFFIX                                                                                         
 PLUGINS                          ALL                                                               
 TARGET_CMAKE_FOLDER              lib/cmake/Elektra                                                 
 TARGET_DOCUMENTATION_DOC-BASE_   share/doc-base/                                                   
 TARGET_DOCUMENTATION_HTML_FOLD   share/doc/elektra-api/html                                        
 TARGET_DOCUMENTATION_LATEX_FOL   share/doc/elektra-api/latex                                       
 TARGET_DOCUMENTATION_MAN_FOLDE   share/man/man3                                                    
 TARGET_DOCUMENTATION_TEXT_FOLD   share/doc/elektra                                                 
 TARGET_INCLUDE_FOLDER            elektra                                                           
 TARGET_PKGCONFIG_FOLDER          pkgconfig                                                         
 TARGET_PLUGIN_FOLDER             elektra                                                           
 TARGET_TEMPLATE_FOLDER           share/elektra/templates                                           

BINDINGS: Which bindings should be added? ALL for all available, DEFAULT for minimal set, see doc/COM
Press [enter] to edit option Press [d] to delete an entry                        CMake Version 3.16.3
Press [c] to configure
Press [h] for help           Press [q] to quit without generating
Press [t] to toggle advanced mode (Currently Off)
```

**Comment by kodebach** [comment]
> ```shell
> No plugin that provides xml could be found
> ```

You can check for plugins that support `xml` with:

```sh
kdb plugin-list xml
```

This will only list plugins that are installed on your system. To find out which plugins exist in the source tree, you can look at the `README.md` files of all plugins:

```sh
grep '^- infos/provides = .*storage/xml' src/plugins/*/README.md
#> src/plugins/xerces/README.md:- infos/provides = storage/xml
#> src/plugins/xmltool/README.md:- infos/provides = storage/xml
```

Both these plugins have external dependencies, so they are excluded from the `PLUGINS=ALL` option if a dependency can't be found. You'll see something like

```
Exclude plugin xerces because XercesC library libxerces-c-dev not found
```

in the output of `cmake` to inform you of this.

In short, you can either install `libxerces-c-dev` and use the [`xerces`](https://www.libelektra.org/plugins/xerces) plugin or install `libxml2-dev` and use [`xmltool`](https://www.libelektra.org/plugins/xmltool). (Note: the package names are generally for Debian, they may be different for your distro)

Both plugins have pros and cons. For info please take a look at the respective READMEs.

**Comment by Kochise** [comment]
Thank for the feedback.

I just want an xml plugin that do it's job.

The best would have metadata as node attributes, I guess.

**Comment by markus2330** [comment]
> I just want an xml plugin that do it's job.
> The best would have metadata as node attributes, I guess.

xerces should do what you want

**Comment by markus2330** [comment]
Please open a new issue if there are any further questions! Feedback on how we can improve the user experience is also highly welcomed, e.g., which kind of tutorial would help you in what you try to do with Elektra.

**Comment by Kochise** [comment]
I reopen the issue because while your answer was pretty clear and I can mount a XML file, I cannot at `user:` level :

````bash
/workdir # wget https://www.w3schools.com/xml/note.xml
Connecting to www.w3schools.com (192.229.133.221:443)
saving to 'note.xml'
note.xml             100% |****************************************************|   164  0:00:00 ETA
'note.xml' saved
/workdir # cat note.xml 
<?xml version="1.0" encoding="UTF-8"?>
<note>
  <to>Tove</to>
  <from>Jani</from>
  <heading>Reminder</heading>
  <body>Don't forget me this weekend!</body>
/workdir # kdb ls system:/elektra/mountpoints
/workdir # kdb mount -W `pwd`/note.xml user:/sw/myorg/toto/#0/current/ xml
Note that absolute paths are still relative to their namespace (see `kdb plugin-info resolver`).
Only system+spec mountpoints are actually absolute.
Use `kdb file user:/sw/myorg/toto/#0/current` to determine where the file(s) are.
Use `-q` or use `kdb set /sw/elektra/kdb/#0/current/quiet 1` to suppress infos.
/workdir # kdb ls system:/elektra/mountpoints
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/config
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/config/path
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/errorplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/errorplugins/#5#resolver_fm_hpu_b#resolver#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins/#0#resolver
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins/#5#xerces#xml#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/mountpoint
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#0#resolver
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#5#xml
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#6#sync#sync#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#7#resolver
/workdir # kdb ls user:/sw/myorg/toto/#0/current/
/workdir # kdb get /sw/myorg/toto/#0/current/heading
Did not find key '/sw/myorg/toto/#0/current/heading'
/workdir # 
```

Yet I **NEED** to mount XML files at different layers, at `system:` to hold default values and CONSTANT, at `user:` to store the application's values, using the cascading read to get either the `user:` value or the `system:` default value.

I don't want to "import" anything.

**Comment by markus2330** [comment]
Thank you for your question!

Yes, as you expect you can do multiple mounts and consider all of them with cascading lookup! If all of them should have the same file name, you can use cascading mounts by leaving the namespace during mounting: `kdb mount -W note.xml system:/sw/myorg/toto/#0/current/ xml`.

The trouble you seem to have is that mountpoints in `user:` always prefix the mounted files with your home directory. This is intentional as the `user:` is designed for user-specific configuration. Of course you could implement your own resolver and do something else but probably this is not necessary:

```
kdb mount -W X/note.xml user:/sw/myorg/toto/#0/current/ xml
```

Where X is the current path but without the prefixed home directory, e.g., if the home directory is /home/kochise and the folder note.xml is in is called "/home/kochise/workdir", you would use:


```
kdb mount -W /workdir/note.xml user:/sw/myorg/toto/#0/current/ xml
```

If it must be the absolute path /workdir for `user:`, please give more details what you want to achive.

**Comment by Kochise** [comment]
Ok, that "makes sense", in a way, that the `user:` configuration is looked for inside the user's folder.

But I think I'm doing something wrong :

```bash
/workdir # ls
/workdir # ls ~
note.xml
/workdir # cat ~/note.xml 
<?xml version="1.0" encoding="UTF-8"?>
<note>
  <to>Tove</to>
  <from>Jani</from>
  <heading>Reminder</heading>
  <body>Don't forget me this weekend!</body>
</note>/workdir # kdb ls system:/elektra/mountpoints
/workdir # kdb mount -W note.xml user:/sw/myorg/toto/#0/current/ xml
/workdir # kdb ls system:/elektra/mountpoints
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/config
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/config/path
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/errorplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/errorplugins/#5#resolver_fm_hpu_b#resolver#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins/#0#resolver
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins/#5#xerces#xml#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/mountpoint
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#0#resolver
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#5#xml
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#6#sync#sync#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#7#resolver
/workdir # kdb ls user:/sw/myorg/toto/#0/current/
/workdir # kdb ls /sw/myorg/toto/#0/current/
/workdir # kdb get /sw/myorg/toto/#0/current/heading
Did not find key '/sw/myorg/toto/#0/current/heading'
/workdir # kdb umount user:/sw/myorg/toto/#0/current/
/workdir # kdb mount -W /note.xml user:/sw/myorg/toto/#0/current/ xml
Note that absolute paths are still relative to their namespace (see `kdb plugin-info resolver`).
Only system+spec mountpoints are actually absolute.
Use `kdb file user:/sw/myorg/toto/#0/current` to determine where the file(s) are.
Use `-q` or use `kdb set /sw/elektra/kdb/#0/current/quiet 1` to suppress infos.
/workdir # kdb umount user:/sw/myorg/toto/#0/current/
/workdir # kdb mount -W ~/note.xml user:/sw/myorg/toto/#0/current/ xml
Note that absolute paths are still relative to their namespace (see `kdb plugin-info resolver`).
Only system+spec mountpoints are actually absolute.
Use `kdb file user:/sw/myorg/toto/#0/current` to determine where the file(s) are.
Use `-q` or use `kdb set /sw/elektra/kdb/#0/current/quiet 1` to suppress infos.
/workdir # kdb umount user:/sw/myorg/toto/#0/current/
/workdir # cd ~
~ # ls
note.xml
~ # kdb ls user:/sw/myorg/toto/#0/current/
~ # kdb ls system:/elektra/mountpoints
~ # kdb mount -W note.xml user:/sw/myorg/toto/#0/current/ xml
~ # kdb ls system:/elektra/mountpoints
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/config
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/config/path
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/errorplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/errorplugins/#5#resolver_fm_hpu_b#resolver#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins/#0#resolver
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/getplugins/#5#xerces#xml#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/mountpoint
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#0#resolver
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#5#xml
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#6#sync#sync#
system:/elektra/mountpoints/user:\/sw\/myorg\/toto\/#0\/current/setplugins/#7#resolver
~ # kdb ls user:/sw/myorg/toto/#0/current/
~ # kdb ls /sw/myorg/toto/#0/current/
~ # kdb get /sw/myorg/toto/#0/current/heading
Did not find key '/sw/myorg/toto/#0/current/heading'
~ # 
```

I don't quite understand.

**Comment by Kochise** [comment]
Then why did this happened :

```bash
~ # ls
note.xml
```
```bash
/workdir # kdb mount -W /note.xml user:/sw/myorg/toto/#0/current/ xml
Note that absolute paths are still relative to their namespace (see `kdb plugin-info resolver`).
Only system+spec mountpoints are actually absolute.
Use `kdb file user:/sw/myorg/toto/#0/current` to determine where the file(s) are.
Use `-q` or use `kdb set /sw/elektra/kdb/#0/current/quiet 1` to suppress infos.
```

How do `elektra` manage to know the system and user paths ?

**Comment by markus2330** [comment]
This is only a note, it should have worked anyway? As written in the output, with `-q` you can get rid of the note.

> How do elektra manage to know the system and user paths ?

Elektra uses CMake variables and APIs (e.g. for home dir). See https://www.libelektra.org/plugins/resolver

**Comment by Kochise** [comment]
Ok, I'm starting to get it, thank you.
